File: 2023/day19/clojure/solution.clj
#!/usr/bin/env clojure -M

(require '[clojure.string :as str])

(defn parse-rule
  "Parse a single rule like 'x>10:one' or 'A' (default)."
  [rule-str]
  (if (str/includes? rule-str ":")
    (let [[condition destination] (str/split rule-str #":")
          attr (keyword (subs condition 0 1))
          op (subs condition 1 2)
          value (parse-long (subs condition 2))]
      {:attr attr :op op :value value :dest destination})
    {:attr nil :op nil :value nil :dest rule-str}))

(defn parse-workflow
  "Parse a workflow line like 'px{a<2006:qkq,m>2090:A,rfg}'."
  [line]
  (let [[name rules-str] (str/split line #"\{")
        rules-str (subs rules-str 0 (dec (count rules-str)))
        rules (mapv parse-rule (str/split rules-str #","))]
    [name rules]))

(defn parse-part
  "Parse a part line like '{x=787,m=2655,a=1222,s=2876}'."
  [line]
  (let [content (subs line 1 (dec (count line)))
        pairs (str/split content #",")]
    (into {}
          (for [pair pairs
                :let [[k v] (str/split pair #"=")]]
            [(keyword k) (parse-long v)]))))

(defn parse-input
  "Parse the input file into workflows map and parts vector."
  [filename]
  (let [text (slurp filename)
        [workflow-section parts-section] (str/split text #"\n\n")
        workflows (into {} (map parse-workflow (str/split-lines workflow-section)))
        parts (mapv parse-part (str/split-lines (str/trim parts-section)))]
    [workflows parts]))

(defn process-part
  "Process a part through workflows, return true if accepted."
  [workflows part]
  (loop [current "in"]
    (cond
      (= current "A") true
      (= current "R") false
      :else
      (let [rules (get workflows current)]
        (recur
         (loop [rules rules]
           (let [{:keys [attr op value dest]} (first rules)]
             (cond
               (nil? attr) dest
               (and (= op "<") (< (get part attr) value)) dest
               (and (= op ">") (> (get part attr) value)) dest
               :else (recur (rest rules))))))))))

(defn part1
  "Sum ratings of accepted parts."
  [workflows parts]
  (->> parts
       (filter #(process-part workflows %))
       (map #(+ (:x %) (:m %) (:a %) (:s %)))
       (reduce + 0)))

(defn count-accepted
  "Count combinations of xmas values that lead to acceptance using range splitting."
  [workflows workflow ranges]
  (cond
    (= workflow "R") 0
    (= workflow "A") (->> (vals ranges)
                          (map (fn [[lo hi]] (max 0 (inc (- hi lo)))))
                          (reduce * 1))
    :else
    (let [rules (get workflows workflow)]
      (loop [rules rules
             ranges ranges
             total 0]
        (if (empty? rules)
          total
          (let [{:keys [attr op value dest]} (first rules)]
            (if (nil? attr)
              ;; Default rule
              (+ total (count-accepted workflows dest ranges))
              ;; Conditional rule
              (let [[lo hi] (get ranges attr)]
                (if (= op "<")
                  ;; Split: [lo, value-1] goes to dest, [value, hi] continues
                  (let [matches? (< lo value)
                        continues? (>= hi value)
                        match-count (if matches?
                                      (count-accepted workflows dest
                                                      (assoc ranges attr [lo (min hi (dec value))]))
                                      0)]
                    (if continues?
                      (recur (rest rules)
                             (assoc ranges attr [(max lo value) hi])
                             (+ total match-count))
                      (+ total match-count)))
                  ;; op == ">"
                  ;; Split: [value+1, hi] goes to dest, [lo, value] continues
                  (let [matches? (> hi value)
                        continues? (<= lo value)
                        match-count (if matches?
                                      (count-accepted workflows dest
                                                      (assoc ranges attr [(max lo (inc value)) hi]))
                                      0)]
                    (if continues?
                      (recur (rest rules)
                             (assoc ranges attr [lo (min hi value)])
                             (+ total match-count))
                      (+ total match-count))))))))))))

(defn part2
  "Count all possible accepted combinations (1-4000 for each rating)."
  [workflows]
  (let [initial-ranges {:x [1 4000]
                        :m [1 4000]
                        :a [1 4000]
                        :s [1 4000]}]
    (count-accepted workflows "in" initial-ranges)))

(defn -main []
  (let [input-file (str (.getParent (java.io.File. *file*)) "/../input.txt")
        [workflows parts] (parse-input input-file)]
    (println "Part 1:" (part1 workflows parts))
    (println "Part 2:" (part2 workflows))))

(-main)

File: 2023/day19/node/solution.js
#!/usr/bin/env node
/**
 * Day 19: Aplenty - Workflow processing and range analysis.
 */

import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseInput(filename) {
  const text = readFileSync(filename, 'utf-8').trim();
  const [workflowSection, partsSection] = text.split('\n\n');

  // Parse workflows
  const workflows = {};
  for (const line of workflowSection.split('\n')) {
    const [name, rulesStr] = line.split('{');
    const rules = rulesStr.slice(0, -1).split(',').map(rule => {
      if (rule.includes(':')) {
        const [condition, destination] = rule.split(':');
        const match = condition.match(/([xmas])([<>])(\d+)/);
        return {
          attr: match[1],
          op: match[2],
          value: parseInt(match[3]),
          destination
        };
      } else {
        return { attr: null, op: null, value: null, destination: rule };
      }
    });
    workflows[name] = rules;
  }

  // Parse parts
  const parts = partsSection.split('\n').map(line => {
    const part = {};
    for (const match of line.matchAll(/([xmas])=(\d+)/g)) {
      part[match[1]] = parseInt(match[2]);
    }
    return part;
  });

  return { workflows, parts };
}

function processPart(workflows, part) {
  let current = 'in';

  while (current !== 'A' && current !== 'R') {
    for (const { attr, op, value, destination } of workflows[current]) {
      if (attr === null) {
        current = destination;
        break;
      } else if (op === '<' && part[attr] < value) {
        current = destination;
        break;
      } else if (op === '>' && part[attr] > value) {
        current = destination;
        break;
      }
    }
  }

  return current === 'A';
}

function part1(workflows, parts) {
  let total = 0;
  for (const part of parts) {
    if (processPart(workflows, part)) {
      total += part.x + part.m + part.a + part.s;
    }
  }
  return total;
}

/**
 * Count combinations of xmas values that lead to acceptance.
 * Uses range splitting to process all possible paths through workflows.
 */
function countAccepted(workflows, workflow, ranges) {
  if (workflow === 'R') {
    return 0n;
  }
  if (workflow === 'A') {
    // Count all combinations in current ranges
    let result = 1n;
    for (const [lo, hi] of Object.values(ranges)) {
      result *= BigInt(Math.max(0, hi - lo + 1));
    }
    return result;
  }

  let total = 0n;
  ranges = { ...ranges };  // Make a copy

  for (const { attr, op, value, destination } of workflows[workflow]) {
    if (attr === null) {
      total += countAccepted(workflows, destination, ranges);
    } else {
      const [lo, hi] = ranges[attr];

      if (op === '<') {
        // Split: [lo, value-1] goes to destination, [value, hi] continues
        if (lo < value) {
          const newRanges = { ...ranges };
          newRanges[attr] = [lo, Math.min(hi, value - 1)];
          total += countAccepted(workflows, destination, newRanges);
        }
        // Remaining part continues to next rule
        if (hi >= value) {
          ranges[attr] = [Math.max(lo, value), hi];
        } else {
          break;
        }
      } else {  // op === '>'
        // Split: [value+1, hi] goes to destination, [lo, value] continues
        if (hi > value) {
          const newRanges = { ...ranges };
          newRanges[attr] = [Math.max(lo, value + 1), hi];
          total += countAccepted(workflows, destination, newRanges);
        }
        // Remaining part continues to next rule
        if (lo <= value) {
          ranges[attr] = [lo, Math.min(hi, value)];
        } else {
          break;
        }
      }
    }
  }

  return total;
}

function part2(workflows) {
  const initialRanges = {
    x: [1, 4000],
    m: [1, 4000],
    a: [1, 4000],
    s: [1, 4000]
  };
  return countAccepted(workflows, 'in', initialRanges);
}

const { workflows, parts } = parseInput(join(__dirname, '..', 'input.txt'));
console.log(`Part 1: ${part1(workflows, parts)}`);
console.log(`Part 2: ${part2(workflows)}`);
